clang/test/SemaHLSL/resource_binding_attr_error_space.hlsl

```diff
@@ -0,0 +1,56 @@
+// RUN: %clang_cc1 -triple dxil-pc-shadermodel6.3-library -x hlsl -o - -fsyntax-only %s -verify
+
+// valid
+cbuffer cbuf {
+    RWBuffer<int> r : register(u0, space0);
+}
+
+cbuffer cbuf2 {
+    struct x {
+        // expected-error@+1 {{'register' attribute only applies to cbuffer/tbuffer and external global variables}}
+        RWBuffer<int> E : register(u2, space3);
+    };
+}
+
+struct MyStruct {
+    RWBuffer<int> E;
+};
+
+cbuffer cbuf3 {
+  MyStruct E : register(u2, space3);

```

Comment:
This will hit error in dxc
```
invalid register specification, expected 'b', 'c', or 'i' binding
```

Reply:
Compilation succeeds with -T lib_6_3 on godbolt. 
I think we expect this to succeed in clang.

Reply:
Sharing the godbolt links to the case where @python3kgae hit the error and @bob80905 doesn't hit the error would probably help the discussion here.

Reply:
https://godbolt.org/z/7axojs5hP

Reply:
https://godbolt.org/z/G7cKdP3hY
cbuffer cbuf3 {
  float a : register(u2, space3);
}

Reply:
> And I think that we said that this would report a "can't use space on globals" error?

I would argue those 2 cases are not equivalent, because one register annotation is not in the global scope, and the other is. So, the second case would emit an error about the annotation not being in the global scope, and the first would not.
I've explicitly added the second case along with the expected error for clarity.

Reply:
> https://godbolt.org/z/G7cKdP3hY cbuffer cbuf3 { float a : register(u2, space3); }

I agree that the code you provided should emit an error, I'll add your example to the test.

Reply:
> > And I think that we said that this would report a "can't use space on globals" error?
> 
> I would argue those 2 cases are not equivalent, because one register annotation is not in the global scope, and the other is. So, the second case would emit an error about the annotation not being in the global scope, and the first would not. I've explicitly added the second case along with the expected error for clarity.

I disagree with this - we could file an issue on this and follow up since this change has other worthy things in it.  Since DXC fails to do the right thing in this case it is arguably more a bug in DXC that it doesn't detect the error here.

Or we could open a can of worms and allow spaces on globals like this, but when we discussed this previously we decided to just match DXC's rules rather than try and make new ones.

Reply:
> https://godbolt.org/z/G7cKdP3hY cbuffer cbuf3 { float a : register(u2, space3); }

That's not the same case that's in the test.  This case has a numeric type, but the test has a UDT.  These have different rules.